algorithms/src/traits/snark.rs

```diff
@@ -26,27 +28,48 @@ pub trait Prepare<T> {
 }
 
 pub trait SNARK {
-    type AllocatedCircuit;
-    type Circuit;
+    type ScalarField: Clone + PrimeField;
+    type BaseField: Clone + PrimeField;
```

Comment:
Explain: Now the SNARK requires specifying the fields. This will help us bound the VerifyingKey to be `ToConstraintField<Self::BaseField>`.